ravenframework/CodeInterfaceClasses/Abce/AbceCodeInterface.py

```diff
@@ -0,0 +1,169 @@
+    for inf in inputFiles:
+      ext = '.' + inf.getExt() if inf.getExt() is not None else ''
+      try:
+        usedExts.remove(ext)
+      except ValueError:
+        pass
+    if len(usedExts) != 0:
+      raise IOError('Input extension',','.join(usedExts),'listed in XML node Code, but not found in the list of Input of <Files>')
+
+    def findSettingIndex(inputFiles,ext):
+      """
+      Find the settings file and return its index in the inputFiles list.
+      @ In, inputFiles, list of InputFile objects
+      @ In, ext, string, extension of the settings.yml file
+      @ Out, index, int, index of the settings file in the inputFiles list
+      """
+      for index,inputFile in enumerate(inputFiles):
+        if inputFile.getBase() == 'settings' and inputFile.getExt() == ext:
+          return index,inputFile
+      raise IOError('No settings file with extension '+ext+' found!')
+
+    def setOutputDir(settingsFile):
+      """
+      Set the output directory in the settings file.
+      @ In, settingsFile, InputFile object, settings.yml file
+      @ Out, None
+      """
+      # the settings file is the settings.yml file the scenario name is in node
+      # simulation -> scenario_name
+      # the output directory is the directory of the settings file in subdirectory "outputs/$scenario_name"
+      # get the scenario name by reading the settings.yml file
+      scenarioName = None
+      with open(settingsFile.getAbsFile(),'r') as settings:
+        for line in settings:
+          if 'scenario_name' in line:
+            scenarioName = line.split(':')[1].strip()
+            # remove the double quotes
+            scenarioName = scenarioName.replace('"','')
+            break
+      self._outputDirectory = os.path.join(os.path.dirname(settingsFile.getAbsFile()),'outputs',scenarioName)
+      #print("_outputDirectory", self._outputDirectory, "settingsFile", settingsFile.getAbsFile(), "scenarioName", scenarioName)
+      return None
+
+    #prepend
+    fullCommand = ''
+    fullCommand += clargs['pre']+' '
+    fullCommand += executable
+    index=None
+    #setup input files and output directory
+    self._outputDirectory = None
+    #inputs
+    for flag,elems in clargs['input'].items():
+      if flag == 'noarg':
+        continue
+      fullCommand += ' '+flag
+      for elem in elems:
+        ext, delimiter = elem[0], elem[1]
+        idx,fname = findSettingIndex(inputFiles,ext.strip('.'))
+        setOutputDir(fname)
+        fullCommand += delimiter + fname.getFilename()
+        if index == None:
+          index = idx
+    self.caseName = inputFiles[index].getBase()
+    outFile = 'out~'+self.caseName
+    if 'output' in clargs:
+      fullCommand+=' '+clargs['output']+' '+outFile
+    fullCommand+=' '+clargs['text']
+    fullCommand+=' '+clargs['post']
+    returnCommand = [('parallel',fullCommand)],outFile
+    print('Execution Command: '+str(returnCommand[0]))
+    return returnCommand
```

Comment:
maybe I'm reading this wrong, but what method is this code part of? It appears to be on the class level to me

Reply:
This is in generateCommand. note that setOutputDir is a function local to generateCommand.